### PR TITLE
node: fix shutdown

### LINKED ISSUE #1034: Double panic on Ctrl+C exit during block sync
Compiler: `rustc 1.31.0-nightly (96064eb61 2018-10-28)`

Branch: `bbq-birch`
```
commit 15e955dc652e48f3e1b13a1b6a93e555ad18203a
Author: Tomasz Drwięga <tomusdrw@users.noreply.github.com>
Date:   Tue Oct 16 10:04:19 2018 +0200

    Longevity handling. (#903)

```

If I send SIGINT to the process I may get SIGILL with 20% chance.

```
(lldb) cont
Process 1968 resuming
thread panicked while processing panic. aborting.
Process 1968 stopped
* thread #10: tid = 1978, 0x0000555556571de4 substrate`std::panicking::rust_panic_with_hook::hcb9ad4b0875cf96f + 148 at panicking.rs:0, name = 'ImportQueue', stop reason = signal SIGILL: illegal instruction operand
    frame #0: 0x0000555556571de4 substrate`std::panicking::rust_panic_with_hook::hcb9ad4b0875cf96f + 148 at panicking.rs:0
(lldb) bt
* thread #10: tid = 1978, 0x0000555556571de4 substrate`std::panicking::rust_panic_with_hook::hcb9ad4b0875cf96f + 148 at panicking.rs:0, name = 'ImportQueue', stop reason = signal SIGILL: illegal instruction operand
  * frame #0: 0x0000555556571de4 substrate`std::panicking::rust_panic_with_hook::hcb9ad4b0875cf96f + 148 at panicking.rs:0
    frame #1: 0x0000555556571a62 substrate`std::panicking::continue_panic_fmt::hee1745c3ca8a23be + 162 at panicking.rs:390
    frame #2: 0x0000555556571946 substrate`rust_begin_unwind + 6 at panicking.rs:325
    frame #3: 0x00005555565b2bed substrate`core::panicking::panic_fmt::hbebfd4901a566206 + 61 at panicking.rs:77
    frame #4: 0x00005555565b2c63 substrate`core::option::expect_failed::h6f7158703abd4906 + 115 at option.rs:1008
    frame #5: 0x00005555565653b9 substrate`std::io::stdio::stderr::he8f02dd76dc6470d + 457 at option.rs:322
    frame #6: 0x0000555555c7c90f substrate`substrate_cli::panic_hook::panic_hook::h1e354d3aae01addd + 351 at panic_hook.rs:53
    frame #7: 0x0000555555c7c609 substrate`core::ops::function::Fn::call::hd06a29a872e1c181 + 9 at function.rs:78
    frame #8: 0x0000555556571f17 substrate`std::panicking::rust_panic_with_hook::hcb9ad4b0875cf96f + 455 at panicking.rs:480
    frame #9: 0x0000555556571a62 substrate`std::panicking::continue_panic_fmt::hee1745c3ca8a23be + 162 at panicking.rs:390
    frame #10: 0x0000555556571946 substrate`rust_begin_unwind + 6 at panicking.rs:325
    frame #11: 0x00005555565b2bed substrate`core::panicking::panic_fmt::hbebfd4901a566206 + 61 at panicking.rs:77
    frame #12: 0x00005555565b2c63 substrate`core::option::expect_failed::h6f7158703abd4906 + 115 at option.rs:1008
    frame #13: 0x00005555565653b9 substrate`std::io::stdio::stderr::he8f02dd76dc6470d + 457 at option.rs:322
    frame #14: 0x0000555555c7c90f substrate`substrate_cli::panic_hook::panic_hook::h1e354d3aae01addd + 351 at panic_hook.rs:53
    frame #15: 0x0000555555c7c609 substrate`core::ops::function::Fn::call::hd06a29a872e1c181 + 9 at function.rs:78
    frame #16: 0x0000555556571f17 substrate`std::panicking::rust_panic_with_hook::hcb9ad4b0875cf96f + 455 at panicking.rs:480
    frame #17: 0x000055555639ad38 substrate`std::panicking::begin_panic::ha9bd69c01f3c54a7 + 40 at panicking.rs:410
    frame #18: 0x000055555639a519 substrate`_$LT$slog_scope..NoGlobalLoggerSet$u20$as$u20$slog..Drain$GT$::log::h39fb4c925a1a8cb7 + 25 at lib.rs:123
    frame #19: 0x0000555555b206b0 substrate`_$LT$std..thread..local..LocalKey$LT$T$GT$$GT$::with::h0b442aec4f0882ee + 1120 at lib.rs:1675
    frame #20: 0x00005555559b20aa substrate`_$LT$substrate_client..client..Client$LT$B$C$$u20$E$C$$u20$Block$GT$$GT$::import_block::h5ae173a154b7da57 + 9818 at lib.rs:211
    frame #21: 0x000055555598b68c substrate`_$LT$substrate_client..client..Client$LT$B$C$$u20$E$C$$u20$Block$GT$$u20$as$u20$substrate_network..chain..Client$LT$Block$GT$$GT$::import::hbdd3dc4bc592ee1c + 1948 at chain.rs:90
    frame #22: 0x0000555555a609cd substrate`substrate_network::import_queue::import_thread::hc89fa5b8373db35d + 4701 at import_queue.rs:322
    frame #23: 0x0000555555b844d5 substrate`_ZN3std9panicking3try7do_call17hd5101818db596964E.llvm.12423680275706143713 + 37 at mod.rs:409
    frame #24: 0x0000555556582e0a substrate`__rust_maybe_catch_panic + 26 at lib.rs:102
    frame #25: 0x000055555594d8f6 substrate`_$LT$F$u20$as$u20$alloc..boxed..FnBox$LT$A$GT$$GT$::call_box::h15702d321c654371 + 182 at panicking.rs:289
    frame #26: 0x00005555565767be substrate`std::sys::unix::thread::Thread::new::thread_start::h407e7a774ea93e4b + 142 at boxed.rs:682
    frame #27: 0x00007ffff6d866ba libpthread.so.0`start_thread + 202
    frame #28: 0x00007ffff68a641d libc.so.6`__clone + 109
(lldb)
```

CPU info:
```
processor       : 0
vendor_id       : GenuineIntel
cpu family      : 6
model           : 142
model name      : Intel(R) Core(TM) i5-7200U CPU @ 2.50GHz
stepping        : 9
microcode       : 0x8e
cpu MHz         : 500.009
cache size      : 3072 KB
physical id     : 0
siblings        : 4
core id         : 0
cpu cores       : 2
apicid          : 0
initial apicid  : 0
fpu             : yes
fpu_exception   : yes
cpuid level     : 22
wp              : yes
flags           : fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb invpcid_single pti ssbd ibrs ibpb stibp tpr_shadow vnmi flexpriority ept vpid fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid mpx rdseed adx smap clflushopt intel_pt xsaveopt xsavec xgetbv1 xsaves dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp flush_l1d
bugs            : cpu_meltdown spectre_v1 spectre_v2 spec_store_bypass l1tf
bogomips        : 5424.00
clflush size    : 64
cache_alignment : 64
address sizes   : 39 bits physical, 48 bits virtual
power management:
```

### LINKED ISSUE #1035: Panic "slog-scope: no logger set"
Compiler: `rustc 1.31.0-nightly (96064eb61 2018-10-28)`

Branch: `bbq-birch`

```commit 15e955dc652e48f3e1b13a1b6a93e555ad18203a
Author: Tomasz Drwięga <tomusdrw@users.noreply.github.com>
Date:   Tue Oct 16 10:04:19 2018 +0200

    Longevity handling. (#903)
```

Potentially related to #1034. Found when I repeatedly tried to catch SIGILL by sending SIGINT to `substrate` process.

tldr:
```
Thread 'tokio-runtime-worker-3' panicked at 'slog-scope: No logger set. Use `slog_scope::set_global_logger` or `slog_scope::scope`.', /home/korvin/.cargo/registry/src/github.com-1ecc6299db9ec823/slog-scope-4.0.1/lib.rs:123
```

```
(lldb) cont
Process 1888 resuming
2018-10-29 16:53:59 Syncing  4.0 bps, target=#409729 (16 peers), best: #24947 (0xa64b…9a19)

====================

stack backtrace:
   0:     0x55555647b26d - backtrace::backtrace::libunwind::trace::h80965d524d6121bd
                        at /home/korvin/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.8/src/backtrace/libunwind.rs:53
                         - backtrace::backtrace::trace::hc59dcbd0d5cf2fc5
                        at /home/korvin/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.8/src/backtrace/mod.rs:42
   1:     0x55555647a4a2 - backtrace::capture::Backtrace::new_unresolved::h8f53b48bd29cdfe4
                        at /home/korvin/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.8/src/capture.rs:88
                         - backtrace::capture::Backtrace::new::hd8b57580e2e38362
                        at /home/korvin/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.8/src/capture.rs:63
                         - <backtrace::capture::Backtrace as core::default::Default>::default::h418631236069c7c1
                        at /home/korvin/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.8/src/capture.rs:235
   2:     0x55555647a519 - backtrace::capture::Backtrace::new::hd8b57580e2e38362
   3:     0x555555c7c908 - substrate_cli::panic_hook::panic_hook::h1e354d3aae01addd
                        at core/cli/src/panic_hook.rs:51
   4:     0x555555c7c608 - core::ops::function::Fn::call::hd06a29a872e1c181
                        at /rustc/96064eb61dc703c289fae61bdf90593d3e7b2449/src/libcore/ops/function.rs:78
   5:     0x555556571f16 - std::panicking::rust_panic_with_hook::hcb9ad4b0875cf96f
                        at libstd/panicking.rs:480
   6:     0x55555639ad37 - std::panicking::begin_panic::ha9bd69c01f3c54a7
                        at /rustc/96064eb61dc703c289fae61bdf90593d3e7b2449/src/libstd/panicking.rs:410
   7:     0x55555639a518 - <slog_scope::NoGlobalLoggerSet as slog::Drain>::log::h39fb4c925a1a8cb7
                        at /home/korvin/.cargo/registry/src/github.com-1ecc6299db9ec823/slog-scope-4.0.1/lib.rs:123
   8:     0x555555884a0c - <alloc::sync::Arc<D> as slog::Drain>::log::h1f2014a682a3dd91
                        at /home/korvin/.cargo/registry/src/github.com-1ecc6299db9ec823/slog-2.2.3/src/lib.rs:1675
                         - <slog::Logger<D>>::log::h0ef82393c4ee4923
                        at /home/korvin/.cargo/registry/src/github.com-1ecc6299db9ec823/slog-2.2.3/src/lib.rs:1189
                         - substrate_cli::informant::start::{{closure}}::{{closure}}::h79db74c356171660
                        at /home/korvin/work/Parity/substrate/<::slog::slog_log macros>:4
   9:     0x55555588d8d9 - slog_scope::with_logger::{{closure}}::h50eb6bce5fffe073
                        at /home/korvin/.cargo/registry/src/github.com-1ecc6299db9ec823/slog-scope-4.0.1/lib.rs:215
                         - <std::thread::local::LocalKey<T>>::try_with::h7521c4fd4c294151
                        at /rustc/96064eb61dc703c289fae61bdf90593d3e7b2449/src/libstd/thread/local.rs:300
                         - <std::thread::local::LocalKey<T>>::with::heaee6f56a5297a8e
                        at /rustc/96064eb61dc703c289fae61bdf90593d3e7b2449/src/libstd/thread/local.rs:254
  10:     0x5555558390f5 - slog_scope::with_logger::hf075cddee494f082
                        at /home/korvin/.cargo/registry/src/github.com-1ecc6299db9ec823/slog-scope-4.0.1/lib.rs:211
                         - substrate_cli::informant::start::{{closure}}::h4922b45ec9ca79ea
                        at /home/korvin/work/Parity/substrate/<::substrate_telemetry::telemetry macros>:2
                         - <futures::stream::for_each::ForEach<S, F, U> as futures::future::Future>::poll::ha6d7e646ebf842dc
                        at /home/korvin/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.24/src/stream/for_each.rs:46
  11:     0x5555557db622 - <futures::future::join::MaybeDone<A>>::poll::hd0e076a3b176c070
                        at /home/korvin/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.24/src/future/join.rs:153
                         - <futures::future::join::Join3<A, B, C> as futures::future::Future>::poll::hcf15cf812e8f73fc
                        at /home/korvin/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.24/src/future/join.rs:67
  12:     0x555555808ef8 - <exit_future::UntilExit<F> as futures::future::Future>::poll::h8eaf029e7c727f8f
                        at /home/korvin/.cargo/registry/src/github.com-1ecc6299db9ec823/exit-future-0.1.2/src/lib.rs:24
                         - <futures::future::map::Map<A, F> as futures::future::Future>::poll::ha9a733c4210dd483
                        at /home/korvin/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.24/src/future/map.rs:30
  13:     0x55555653f3a1 - <alloc::boxed::Box<F> as futures::future::Future>::poll::hb55229b9a2ef3afc
                        at /home/korvin/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.24/src/future/mod.rs:113
                         - <futures::task_impl::Spawn<T>>::poll_future_notify::{{closure}}::ha93b47bceb706735
                        at /home/korvin/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.24/src/task_impl/mod.rs:314
                         - <futures::task_impl::Spawn<T>>::enter::{{closure}}::h81108bce414f26f6
                        at /home/korvin/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.24/src/task_impl/mod.rs:388
                         - futures::task_impl::std::set::h2fed1a0a8a6d05b6
                        at /home/korvin/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.24/src/task_impl/std/mod.rs:78
  14:     0x555556542841 - <futures::task_impl::Spawn<T>>::enter::h0aaaa45ba2bae4f2
                        at /home/korvin/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.24/src/task_impl/mod.rs:388
                         - <futures::task_impl::Spawn<T>>::poll_future_notify::hcaaacd4d55503faf
                        at /home/korvin/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-0.1.24/src/task_impl/mod.rs:314
                         - tokio_threadpool::task::TaskFuture::poll::h76814d3119c8d1c0
                        at /home/korvin/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-threadpool-0.1.4/src/task/mod.rs:292
                         - tokio_threadpool::task::Task::run::{{closure}}::h4fcaf0f145d4dada
                        at /home/korvin/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-threadpool-0.1.4/src/task/mod.rs:165
                         - core::ops::function::FnOnce::call_once::h40d38b5c80e6c187
                        at /rustc/96064eb61dc703c289fae61bdf90593d3e7b2449/src/libcore/ops/function.rs:238
                         - <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once::h1c97e1f3264d744a
                        at /rustc/96064eb61dc703c289fae61bdf90593d3e7b2449/src/libstd/panic.rs:313
                         - std::panicking::try::do_call::h15dd9e1b40a78b5b
                        at /rustc/96064eb61dc703c289fae61bdf90593d3e7b2449/src/libstd/panicking.rs:310
  15:     0x555556582e09 - __rust_maybe_catch_panic
                        at libpanic_unwind/lib.rs:102
  16:     0x555556541cd8 - std::panicking::try::hc2ffecd0c30c525c
                        at /rustc/96064eb61dc703c289fae61bdf90593d3e7b2449/src/libstd/panicking.rs:289
                         - std::panic::catch_unwind::h903c8313d8ce6b66
                        at /rustc/96064eb61dc703c289fae61bdf90593d3e7b2449/src/libstd/panic.rs:392
                         - tokio_threadpool::task::Task::run::h69582c81c99b88f7
                        at /home/korvin/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-threadpool-0.1.4/src/task/mod.rs:151
  17:     0x55555653b725 - tokio_threadpool::worker::Worker::run_task2::hab301107a1503377
                        at /home/korvin/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-threadpool-0.1.4/src/worker/mod.rs:520
                         - tokio_threadpool::worker::Worker::run_task::hbbc6a7333a195806
                        at /home/korvin/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-threadpool-0.1.4/src/worker/mod.rs:420
  18:     0x55555653a974 - tokio_threadpool::worker::Worker::try_run_owned_task::h16c67a708dabea5d
                        at /home/korvin/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-threadpool-0.1.4/src/worker/mod.rs:360
                         - tokio_threadpool::worker::Worker::try_run_task::h10dec3d1774a15ff
                        at /home/korvin/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-threadpool-0.1.4/src/worker/mod.rs:289
                         - tokio_threadpool::worker::Worker::run::h0fe64c9fb71404a2
                        at /home/korvin/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-threadpool-0.1.4/src/worker/mod.rs:235
  19:     0x5555565259e5 - tokio::runtime::builder::Builder::build::{{closure}}::{{closure}}::{{closure}}::{{closure}}::h188baee9aaba5fb1
                        at /home/korvin/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.11/src/runtime/builder.rs:238
                         - tokio_timer::timer::handle::with_default::{{closure}}::h10615c4e9807fb47
                        at /home/korvin/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-timer-0.2.6/src/timer/handle.rs:94
                         - <std::thread::local::LocalKey<T>>::try_with::h5c0bbaeaa150385c
                        at /rustc/96064eb61dc703c289fae61bdf90593d3e7b2449/src/libstd/thread/local.rs:300
                         - <std::thread::local::LocalKey<T>>::with::h4b4d42e2b3b7aea0
                        at /rustc/96064eb61dc703c289fae61bdf90593d3e7b2449/src/libstd/thread/local.rs:254
  20:     0x555556525531 - tokio_timer::timer::handle::with_default::h1cf688c04ca5fac4
                        at /home/korvin/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-timer-0.2.6/src/timer/handle.rs:81
                         - tokio::runtime::builder::Builder::build::{{closure}}::{{closure}}::{{closure}}::ha358191103e361d2
                        at /home/korvin/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.11/src/runtime/builder.rs:237
                         - tokio_timer::clock::clock::with_default::{{closure}}::h0f505625d76f9691
                        at /home/korvin/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-timer-0.2.6/src/clock/clock.rs:136
                         - <std::thread::local::LocalKey<T>>::try_with::h2bad441cdf85d46a
                        at /rustc/96064eb61dc703c289fae61bdf90593d3e7b2449/src/libstd/thread/local.rs:300
                         - <std::thread::local::LocalKey<T>>::with::h1cc3890d3e4fc7d6
                        at /rustc/96064eb61dc703c289fae61bdf90593d3e7b2449/src/libstd/thread/local.rs:254
  21:     0x55555652571b - tokio_timer::clock::clock::with_default::hd2c90f24dd55d34e
                        at /home/korvin/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-timer-0.2.6/src/clock/clock.rs:119
                         - tokio::runtime::builder::Builder::build::{{closure}}::{{closure}}::h17d18414df9cc824
                        at /home/korvin/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.11/src/runtime/builder.rs:236
                         - tokio_reactor::with_default::{{closure}}::h5cf49575ec979527
                        at /home/korvin/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-reactor-0.1.1/src/lib.rs:207
                         - <std::thread::local::LocalKey<T>>::try_with::he8af5c3c0b1b99ae
                        at /rustc/96064eb61dc703c289fae61bdf90593d3e7b2449/src/libstd/thread/local.rs:300
                         - <std::thread::local::LocalKey<T>>::with::h2b627d84fa6661f8
                        at /rustc/96064eb61dc703c289fae61bdf90593d3e7b2449/src/libstd/thread/local.rs:254
  22:     0x55555652b605 - tokio_reactor::with_default::h6503b48439c24aa5
                        at /home/korvin/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-reactor-0.1.1/src/lib.rs:199
                         - tokio::runtime::builder::Builder::build::{{closure}}::h7498a654fabe95dc
                        at /home/korvin/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.1.11/src/runtime/builder.rs:235
  23:     0x5555565412bc - tokio_threadpool::callback::Callback::call::ha37f0e8c8933a969
                        at /home/korvin/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-threadpool-0.1.4/src/callback.rs:21
                         - tokio_threadpool::worker::Worker::do_run::{{closure}}::{{closure}}::hfdccdd9f450af8ec
                        at /home/korvin/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-threadpool-0.1.4/src/worker/mod.rs:117
                         - tokio_executor::global::with_default::{{closure}}::h1703b2fb935c5ae2
                        at /home/korvin/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-executor-0.1.5/src/global.rs:192
                         - <std::thread::local::LocalKey<T>>::try_with::he006b2075fe0d8f7
                        at /rustc/96064eb61dc703c289fae61bdf90593d3e7b2449/src/libstd/thread/local.rs:300
                         - <std::thread::local::LocalKey<T>>::with::h5dd6f5b6539592d9
                        at /rustc/96064eb61dc703c289fae61bdf90593d3e7b2449/src/libstd/thread/local.rs:254
  24:     0x555556541421 - tokio_executor::global::with_default::hef0c548bbcceefe0
                        at /home/korvin/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-executor-0.1.5/src/global.rs:162
                         - tokio_threadpool::worker::Worker::do_run::{{closure}}::h751a1fdf8b23ec73
                        at /home/korvin/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-threadpool-0.1.4/src/worker/mod.rs:115
                         - <std::thread::local::LocalKey<T>>::try_with::h531f3043231ca5f5
                        at /rustc/96064eb61dc703c289fae61bdf90593d3e7b2449/src/libstd/thread/local.rs:300
                         - <std::thread::local::LocalKey<T>>::with::h6ad87a9e9de9b3e5
                        at /rustc/96064eb61dc703c289fae61bdf90593d3e7b2449/src/libstd/thread/local.rs:254
  25:     0x55555654206f - tokio_threadpool::worker::Worker::do_run::h845554f3c99e7c83
                        at /home/korvin/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-threadpool-0.1.4/src/worker/mod.rs:106
                         - tokio_threadpool::pool::Pool::spawn_thread::{{closure}}::hbeff55c4b9f76950
                        at /home/korvin/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-threadpool-0.1.4/src/pool/mod.rs:411
                         - std::sys_common::backtrace::__rust_begin_short_backtrace::h4f0250a737bcb3ae
                        at /rustc/96064eb61dc703c289fae61bdf90593d3e7b2449/src/libstd/sys_common/backtrace.rs:136
  26:     0x55555654288b - std::thread::Builder::spawn::{{closure}}::{{closure}}::hb9ec43d91b77b998
                        at /rustc/96064eb61dc703c289fae61bdf90593d3e7b2449/src/libstd/thread/mod.rs:409
                         - <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once::h505a0de9b5e0dc5f
                        at /rustc/96064eb61dc703c289fae61bdf90593d3e7b2449/src/libstd/panic.rs:313
                         - std::panicking::try::do_call::ha38dee7b3547d193
                        at /rustc/96064eb61dc703c289fae61bdf90593d3e7b2449/src/libstd/panicking.rs:310
  27:     0x555556582e09 - __rust_maybe_catch_panic
                        at libpanic_unwind/lib.rs:102
  28:     0x55555653e56f - std::panicking::try::hd67c1e0934d7184a
                        at /rustc/96064eb61dc703c289fae61bdf90593d3e7b2449/src/libstd/panicking.rs:289
                         - std::panic::catch_unwind::h82d9f1af2cf1cdec
                        at /rustc/96064eb61dc703c289fae61bdf90593d3e7b2449/src/libstd/panic.rs:392
                         - std::thread::Builder::spawn::{{closure}}::h69804e3cb27168a3
                        at /rustc/96064eb61dc703c289fae61bdf90593d3e7b2449/src/libstd/thread/mod.rs:408
                         - <F as alloc::boxed::FnBox<A>>::call_box::he8cd0000be40e897
                        at /rustc/96064eb61dc703c289fae61bdf90593d3e7b2449/src/liballoc/boxed.rs:672
  29:     0x5555565767bd - <alloc::boxed::Box<(dyn alloc::boxed::FnBox<A, Output$u3d$R$GT$$u20$$u2b$$u20$$u27$a$RP$$GT$$u20$as$u20$core..ops..function..FnOnce$LT$A$GT$$GT$::call_once::h2a20cbef9f6dfce7
                        at /rustc/96064eb61dc703c289fae61bdf90593d3e7b2449/src/liballoc/boxed.rs:682
                         - std::sys_common::thread::start_thread::h38c7375c4ca4e476
                        at libstd/sys_common/thread.rs:24
                         - std::sys::unix::thread::Thread::new::thread_start::h407e7a774ea93e4b
                        at libstd/sys/unix/thread.rs:90
  30:     0x7ffff6d866b9 - start_thread
  31:     0x7ffff68a641c - clone
  32:                0x0 - <unknown>

Thread 'tokio-runtime-worker-3' panicked at 'slog-scope: No logger set. Use `slog_scope::set_global_logger` or `slog_scope::scope`.', /home/korvin/.cargo/registry/src/github.com-1ecc6299db9ec823/slog-scope-4.0.1/lib.rs:123

This is a bug. Please report it at:

    https://github.com/paritytech/polkadot/issues/new

Process 1888 exited with status = 1 (0x00000001) 
```

CPU info:
```
processor       : 0
vendor_id       : GenuineIntel
cpu family      : 6
model           : 142
model name      : Intel(R) Core(TM) i5-7200U CPU @ 2.50GHz
stepping        : 9
microcode       : 0x8e
cpu MHz         : 500.009
cache size      : 3072 KB
physical id     : 0
siblings        : 4
core id         : 0
cpu cores       : 2
apicid          : 0
initial apicid  : 0
fpu             : yes
fpu_exception   : yes
cpuid level     : 22
wp              : yes
flags           : fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb invpcid_single pti ssbd ibrs ibpb stibp tpr_shadow vnmi flexpriority ept vpid fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid mpx rdseed adx smap clflushopt intel_pt xsaveopt xsavec xgetbv1 xsaves dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp flush_l1d
bugs            : cpu_meltdown spectre_v1 spectre_v2 spec_store_bypass l1tf
bogomips        : 5424.00
clflush size    : 64
cache_alignment : 64
address sizes   : 39 bits physical, 48 bits virtual
power management:
```

### LINKED ISSUE #1161: Ctrl-C didn't work
When the substrate node running long time, Ctrl-C didn't work.
version: the last master.

### PR DESCRIPTION
**Pull request by andresilva**
Tokio doesn't run destructors for any futures still in the pool on shutdown. We blocked the tokio runtime on the outer exit future (triggered by ctrl-c), afterwards we would fire the exit signal but would not wait for all futures in the runtime to finish. Most significantly this can lead to database corruption since it isn't closed properly. 

Now, after firing the exit signal we wait for the runtime to be idle, which means that all spawned futures must be guarded by the exit signal (aura and grandpa were missing).

Merge after #1306.
Fixes #1161, fixes #1034, fixes #1035.

### CODE REVIEW

File: node/cli/src/lib.rs [hunk 1]
```diff
@@ -172,5 +162,14 @@ fn run_until_exit<T, C, E>(
 
 	let _ = runtime.block_on(e.into_exit());
 	exit_send.fire();
+
+	// we eagerly drop the service so that the internal exit future is fired,
+	// but we need to keep holding a reference to the global telemetry guard
+	let _telemetry = service.telemetry();
+	drop(service);
+
+	// TODO [andre]: timeout this future
```
Comment (bkchr) [inline node/cli/src/lib.rs]: Please make a github issue for this.

File: node/cli/src/service.rs [hunk 2]
```diff
@@ -120,7 +109,7 @@ construct_service_factory! {
 					grandpa::NetworkBridge::new(service.network()),
 				)?;
 
-				executor.spawn(voter);
+				executor.spawn(voter.select(service.on_exit()).then(|_| Ok(())));
```
Comment (bkchr) [inline node/cli/src/service.rs]: Can we not take the `on_exit` as parameter for `run_grandpa`? I think that makes it more easy to use and people who write their own node does not forget to select on exit.

File: node/cli/src/service.rs [hunk 3]
```diff
@@ -77,38 +73,31 @@ construct_service_factory! {
...
+					let aura = start_aura(
+						SlotDuration::get_or_compute(&*client)?,
+						key.clone(),
+						client,
+						block_import.clone(),
+						proposer,
+						service.network(),
+					);
+
+					executor.spawn(aura.select(service.on_exit()).then(|_| Ok(())));
```
Comment (bkchr) [inline node/cli/src/service.rs]: Same here as below.

File: node/cli/src/lib.rs (hunk 1 above)
Comment (andresilva) [reply node/cli/src/lib.rs]: Done.

File: node/cli/src/service.rs (hunk 2 above)
Comment (andresilva) [reply node/cli/src/service.rs]: I've updated the code according to your suggestions, but FWIW I think I prefer the original version  because it's easier to check, we just make sure that all call-sites of `spawn` are exit-guarded. I think longer-term we should go with @rphmeier's suggestion though.
